### DISCUSSION TITLE
[mdx-loader] How to use HTML in markdown (original: how to bypass web components in markdown files?)

### DISCUSSION THREAD

**Discussion by JounQin** [mdx-js/mdx#2114] [Q&A] [answered]
There is only document about Web components in MDX: https://johno.com/web-components-in-mdx

When I developed a web component named `onii-quil`, I want to use it in markdown:

```md
<onii-quil></onii-quil>
```

It renders as `<p></p>` unexpectedly.

I haven't found any solution to bypass `<onii-quil></onii-quil>` as-is.

**Comment by ChristianMurphy** [comment]
From the discussion in https://github.com/mdx-js/mdx/issues/2100
I gather that passing a `components` prop with `{"onii-quil": "onii-quil"}`
Should work.

**Reply by JounQin** [reply]
I'm still confusing where should be

> passing a `components` prop with `{"onii-quil": "onii-quil"}`

Maybe you mean https://mdxjs.com/table-of-components/#components?

I tried with 

```jsx
<Readme components={{ 'onii-quill': 'onii-quill' }} /> 
```

```md
<!-- READMD.md -->

## Demo

<onii-quill />
```

But it seems not working.

`Readme` is a `markdown` file, not `mdx`.

**Reply by wooorm** [reply]
> I gather that passing a components prop with `{"onii-quil": "onii-quil"}`

That discussion is about how such code was internally generated wrongly before, leading to a crash. It is now generating that code properly.
But it’s not needed to pass this as a user!

**Comment by wooorm** [comment, marked as answer]
`<onii-quill />` is not HTML. It’s JSX. If you have markdown, markdown supports HTML, and it won’t support JSX.

For `<onii-quil></onii-quil>`, CommonMark describes your behavior: https://spec.commonmark.org/dingus/?text=%3Conii-quil%3E%3C%2Fonii-quil%3E. If that is what the loader results in, it is correct.
If you want to write it so that the paragraph doesn’t show up, put the opening and closing tag on two lines to get “condition 7” (what I call “complete HTML”) https://spec.commonmark.org/0.30/#html-blocks.
I have in the past found this condition a bit too strict. I can imagine, like you, that a closing tag should be allowed to follow the opening tag. You could raise an issue with CommonMark about this.

If things are disappearing, well, they shouldn’t. Please provide more info so I can help debug your code.

**Reply by JounQin** [reply]
@wooorm 

Here you are: https://github.com/oniijs/onii/tree/repro/web_components

It works as jsx in React:

https://github.com/oniijs/onii/blob/f5cf6a90f844162e5aacff572baa15ac3f1b3928/docs/App.tsx#L64

But not as HTML in markdown:

https://github.com/oniijs/onii/blob/f5cf6a90f844162e5aacff572baa15ac3f1b3928/packages/%40oniijs/quill/README.md?plain=1#L41-L42

---

Step: clone and run `yarn dev`, and visit http://localhost:8080/packages/@oniijs/quill, no content in `Demo` section.

**Reply by JounQin** [reply]
Maybe related to https://github.com/syntax-tree/mdast-util-to-hast/blob/f6a6d07ab3bbc8e3ccd4ea19bfdbeed5af1ffbf8/lib/handlers/html.js#L15?

**Reply by wooorm** [reply]
You’re trying to embed strings of HTML and have them rendered as virtual react components.
That doesn’t work.
You can’t do `React.createElement('<div><!--comment-->whatever')`.
It’s unrelated to custom elements. It’s just: strings of HTML.

You need a full blown HTML parser to turn strings of HTML into something React understands.

See https://github.com/mdx-js/mdx/issues/1820#issuecomment-970430877.

**Reply by JounQin** [reply]
You mean `<onii-quil></onii-quil>` is not parsed as HTML, right? (It's in a markdown file) It seems very unnatural to me personally.

**Reply by wooorm** [reply]
HTML in markdown is a black box. It is invalid HTML. It is XML-like garbage. It’s impossible to “parse” and pass to React.

**Reply by JounQin** [reply]
By adding `rehypePlugins: [(yield import('rehype-raw')).default, {passThrough: nodeTypes}]`, it works!!

**Reply by JounQin** [reply]
I think this is worth to document like https://johno.com/web-components-in-mdx

**Reply by wooorm** [reply]
Yeah. Maybe. Not a lot of people are using MDX with plain markdown yet. Particularly then with HTML in markdown.
And MDX is kind of the better alternative for this. Your using MDX but not using all the magic it adds, instead, you’re now duplicating the code with a different parser.
I’m not sure about when this is a great idea

**Reply by JounQin** [reply]
> And MDX is kind of the better alternative for this. Your using MDX but not using all the magic it adds, instead, you’re now duplicating the code with a different parser.
> I’m not sure about when this is a great idea

I'd prefer README.md instead of README.mdx while can be read easily on GitHub, although custom components will be hidden in this case.
